Exception handling in registry-related code

dgmcdona opened the pull request:
This adds exception handling in a number of places where `InvalidAddressException`s were not being caught before.

Notably, `CM_KEY_VALUE.decode_data()` was missing exception handling for calls to `read` that can raise these exceptions. The new behavior is that a debug message will be logged, and null bytes of the same length will be returned in place of the actual data. This will allow partial data to be returned when reading large binary blobs from binary registry values.

It also adds exception handling in the amcache and hashdump plugins where the exceptions raise by `get_key` were not being caught exhaustively.

dgmcdona committed:
Windows Cachedump: Handle uncaught InvalidAddressException

dgmcdona committed:
Windows Registry: Handle uncaught exceptions

A number of calls to `get_key` across multiple plugins are not made
within a `try/except` block that handles
`registry.RegistryFormatException` - the calls are either unprotected or
only check for `KeyError`. This adds the required `try/except` blocks,
or updates the existing ones as needed.

dgmcdona committed:
Windows Registry: Handle possible exception in get_node

Encountered a `SwappedInvalidAddressException` within the call to `cast`
due to an underlying call to `read`.

dgmcdona committed:
Windows Registry: Handle exceptions in read calls

These calls to `.read()` can raise an `InvalidAddressException`. Instead
of propagating this exception to the caller, this adds debug logging,
and pads the data will null bytes.

Also updates the docstring for `decode_data()` to indicate that it can
raise `TypeError` and `ValueError`.

dgmcdona committed:
Windows Registry: Update docstrings + exceptions

This updates the docstrings on several methods to indicate that they may
raise an exception.

dgmcdona commented on volatility3/framework/symbols/windows/extensions/registry.py:
```diff
@@ -159,14 +159,22 @@ class CM_KEY_NODE(objects.StructType):
+        """
+        Returns a bool indicating whether or not the key is volatile.
+
+        Raises ValueError if the key was not instantiated on a RegistryHive layer
```
Unfortunately this raises `ValueError` when the other methods raise `TypeError` for the same reason, but I left it as is to avoid the major version bump that I suspect would be required if we changed this.

ikelos replied:
Errr, I'm actually willing to go for gray area.  Yes, technically we probably should do a major bump, but equally I'm not sure how many things would break because we changed an exception (the process would already have been throwing an exception).   We might want to check anywhere in our code (at least) where we explicitly catch a ValueError to check that it isn't important.  If so, I'd be willing to let it slide with a MINOR bump rather than a full MAJOR one.  (Not relaxing the rule in general, just making this specific single exception)...

dgmcdona replied:
Ok cool, I did the conversion + version bump, with an explanation for the version bump in the commit message.

ikelos committed:
Merge branch 'develop' into issue_1469_cachedump_exception

dgmcdona committed:
Convert ValueError to TypeError

All other methods in this class raise a `TypeError` if the hive was not
instantiated on a registry layer; this changes makes this method
consistent with the convention used in the others.

All `except` blocks checking for `ValueError` have been audited to
ensure that this doesn't break exception handling in existing code
within the framework. This also includes a minor version bump because:

1. RegistryHives are currently only instantiated one way, which is
   through the `hivelist` plugin. `hivelist` uses the correct layers when
   instantiating the hives.
2. Because there is currently a single source for registry hives, and
   it's unlikely that a hive from that source will ever be created on
   the wrong layer, it's unlikely that the existing `ValueError` is
   being raised anywhere within the framework's code.
3. It seems unlikely that consumers of this framework would be
   instantiating registry hives independent of the `hivelist` plugin,
   given that they would effectively have to duplicate the `hivelist`
   code to do so.

For these reasons, we're going to do a minor version bump, even though
an argument can be made that this warrants a major version bump
according to the SemVer rules. This is a one-off and does not indicate
any change in the way that we typically update version numbers.

dgmcdona commented:
closes #1469

ikelos commented:
Maybe it only does it if it's in the first comment?  Make it's case sensitive?  5:S  It does work usually, honest!

ikelos commented:
Ok, thanks, let's get this in...  5:)
